modules/core/src/main/java/org/apache/ignite/internal/util/GridHandleTable.java

```diff
@@ -107,6 +121,14 @@ public int lookup(Object obj) {
      * Resets table to its initial (empty) state.
      */
     public void clear() {
+        if (size < objs.length) {
+            if (shrink()) {
+                size = 0;
+
+                return;
+            }
+        }
+
         System.arraycopy(spineEmpty, 0, spine, 0, spineEmpty.length);
         System.arraycopy(nextEmpty, 0, next, 0, nextEmpty.length);
```

Comment:
```suggestion
        just some more thoughts: 
-         I guess that the array 'next' can be not cleaned at all as we  are resetting 'size' to zero and we never read the next array elements with index>=size

-       we could consider zero value as empty not -1, if we would change postitioning to 'int idx = (hash(obj) % spine.length) + 1; in this case  we wouldn't need to fill int arrays in init() method as they have zeroes by default.
        
```

Reply:
I like the first suggestion, but the second one will lead to unnecessary complication (because `size` will be affected by this change of indexing) and filling operation shouldn't be very frequent with the fix.

Reply:
I agree with you, let's ignore the second one, it's arguable.